src/manifest/container.rs

```diff
@@ -0,0 +1,490 @@
+//! although the magic number and the manifest body may contain payloads that
+//! are Manticore-specific.
+
+use crate::crypto::rsa;
+use crate::io;
+use crate::io::Cursor;
+use crate::io::Read as _;
+use crate::io::Write;
+use crate::manifest::Error;
+use crate::manifest::ManifestType;
+use crate::protocol::wire::WireEnum;
+
+/// Metadata for a [`Container`].
+///
+/// This struct describes metadata attached to every manifest, which makes up
+/// part of the signed component.
+///
+/// [`Comtainer`]: struct.Container.html
+#[derive(Debug)]
+pub struct Metadata {
+    /// The "version" or "manifest ID", a monotonically increasing integer that
+    /// Manticore can use to protect against playback attacks, by refusing to
+    /// load a manifest with a smaller version number.
+    ///
+    /// When minting a new manifest, a signing authority should make sure to
+    /// bump this value.
+    pub version_id: u32,
+}
+
+/// A parsed, verified, manifest container.
+///
+/// This type represents a generic, authenticated manifest. A value of this
+/// type is a witness that authentication via signature was successful; it is
+/// not possible to parse a `Container` without also verifying it.
+///
+/// See the [module-level documentation](index.html) for more information.
+pub struct Container<'m> {
+    manifest_type: ManifestType,
+    metadata: Metadata,
+    body: &'m [u8],
+}
+
+/// Offsets for fields within the container header.
+const LEN_OFFSET: usize = 0;
+const TYPE_OFFSET: usize = 2;
+const ID_OFFSET: usize = 4;
+const SIG_LEN_OFFSET: usize = 8;
+
+/// The length of the container header in bytes:
+/// two halves, a word, another half, and two bytes of padding.
+const HEADER_LEN: usize = 2 + 2 + 4 + 2 + 2;
+
+impl<'m> Container<'m> {
+    /// Parses and verifies a `Container` using the provided [`rsa::Engine`].
+    ///
+    /// This function first parses the `Container`'s header, which it uses for
+    /// finding the signature at the end of the buffer.
+    ///
+    /// `buf` must be aligned to a four-byte boundary.
+    ///
+    /// [`rsa::Engine`]: ../../crypto/rsa/trait.Engine.html
+    pub fn parse_and_verify<Rsa: rsa::Engine>(
+        buf: &'m [u8],
+        rsa: &mut Rsa,
+    ) -> Result<Self, Error> {
+        if buf.as_ptr().align_offset(4) != 0 {
+            return Err(Error::Unaligned);
+        }
+
+        if HEADER_LEN > buf.len() {
+            return Err(Error::OutOfRange);
+        }
+
+        let mut r = buf; // Use io::Read.
+        let len = r.read_le::<u16>()? as usize;
+        let magic = r.read_le::<u16>()?;
+        let id = r.read_le::<u32>()?;
+        let sig_len = r.read_le::<u16>()? as usize;
+
+        // This length check, combined with the checked arithmetic below,
+        // ensures that none of the slice index operations can panic.
+        if len > buf.len() || len < HEADER_LEN {
+            return Err(Error::OutOfRange);
+        }
+        let rest = &buf[..len][HEADER_LEN..];
```

Comment:
It would be nice to have a comment here that says this takes care of 2 bytes of padding. What do you think?

Reply:
Done.